src/model/guild/guild_id.rs

```diff
@@ -757,3 +784,83 @@ impl<'a> From<&'a Guild> for GuildId {
     /// Gets the Id of Guild.
     fn from(live_guild: &Guild) -> GuildId { live_guild.id }
 }
+
+/// A helper class returned by [`GuildId.members_iter()`]
+///
+/// [`GuildId.members_iter()`]: #method.members_iter
+#[derive(Clone, Debug)]
+#[cfg(all(feature = "http", feature = "cache"))]
+pub struct MembersIter<H: AsRef<Http>> {
+    guild_id: GuildId,
+    http: H,
+    buffer: Vec<Member>,
+    after: Option<UserId>,
+    tried_fetch: bool,
+}
+
+#[cfg(all(feature = "http", feature = "cache"))]
+impl<H: AsRef<Http>> MembersIter<H> {
+    fn new(guild_id: GuildId, http: H) -> MembersIter<H> {
+        MembersIter {
+            guild_id,
+            http: http,
+            buffer: Vec::new(),
+            after: None,
+            tried_fetch: false,
+        }
+    }
+
+    /// Fills the `self.buffer` cache of Members.
+    ///
+    /// This drops any members that
+    /// were currently in the buffer, so it should only be called when
+    /// `self.buffer` is empty.  Additionally, this updates `self.after` so that
+    /// the next call does not return duplicate items.  If there are no more
+    /// members to be fetched, then this marks `self.after` as None, indicating
+    /// that no more calls ought to be made.
+    fn refresh(&mut self) -> Result<()> {
+        // Number of profiles to fetch
+        let grab_size: u64 = 1000;
+
+        let next_members = self.guild_id
+            ._members(self.http.as_ref(), Some(grab_size), self.after)?;
+
+         //Get the last member.  If shorter than 1000, there are no more results anyway
+        self.after = next_members.get(grab_size as usize - 1usize)
+            .map(|member| member.user_id());
+
+        // Reverse to optimize pop()
+        self.buffer = next_members.into_iter().rev().collect();
+
+        self.tried_fetch = true;
+
+        Ok(())
+    }
+}
+
+#[cfg(all(feature = "http", feature = "cache"))]
+impl<H: AsRef<Http>> Iterator for MembersIter<H> {
+    type Item = Result<Member>;
+
+    fn next(&mut self) -> Option<Result<Member>> {
+        if self.buffer.len() == 0 && self.after.is_some() || !self.tried_fetch {
+            if let Err(e) = self.refresh() {
+                return Some(Err(e))
+            }
+        }
+
+        self.buffer.pop().map(|i| Ok(i))
```

Comment:
The `Ok` variant is a function of its own.

```suggestion
        self.buffer.pop().map(Ok)
```